### ISSUE TITLE
Question: Optimizer, choice of learning rate

### ISSUE THREAD

**Issue by jlause** [lmb-freiburg/Unet-Segmentation#35]
Hey,

I am currently experimenting with training a detection U-Net from scratch, and have tried different settings for the learning rate. I saw in the `solver_prototxt` information in the `*modeldef.h5` files generated by the finetuning, that the default optimizer is Adam, with parameters as follows:
```
 base_lr:       0.00001
 momentum:      0.9
 momentum2:     0.999
 lr_policy:     "fixed"
```

 Now, I have a couple of questions to that:

-Is it correct that learning rate set with the ImageJ plugin overwrites the `base_lr=10^-5` setting?
-Do I understand Adam correctly that it adjusts the learning rate per parameter, and the learning rate set in the plugin is just like an upper bound to the optimization stepsize? Or does the `lr_policy="fixed"` imply that the effective stepsize is constant despite the optimzer being Adam?
-Is there any best practice to set or tune the learning rate for U-Nets? Does it e.g. make sense to train from scratch for few iterations with high learning rate, then finetune with smaller learning rate? (I would assume that Adam takes care of this implicitly)

Two more general, related questions:

-I saw in your paper that you trained one of your from-scratch models with SGD and learning rate decay. To implement that with the plugin, is it sufficient to change the `solver_prototxt` attribute of the `*_modeldef.h5` file accordingly?
-I also saw in your paper description of the training procedure that you trained Adam with 'fixed schedule' - what is meant by that? That the `base_lr` is not changed?

I'd be grateful for any hint! Thanks in advance,
Jan.

**Comment by ThorstenFalk** [comment]
Hi Jan,

the plugin interface itself lets you only change the base learning rate (the default is overwritten as you assumed), and yes since ADAM is used it is only sort of an upper bound, and ADAM will change it during optimization. The fixed schedule only means that the base lr will be the same for the whole optimization.

There are no best practices I am aware of. I usually keep the learning rate constant and let Adam do the rest, but as you saw, sometimes a manually selected schedule can lead to better results.
There are some alternatives:
* The step schedule is quite popular, in which you reduce the learning rate by a user-defined factor after a user-defined number of iterations. For the cell net models one could for example use a step schedule with a reduction of factor 2 every 50000 iterations which can be set by modifying the `solver_prototxt` attribute (externally, cannot be done with the plugin at the moment), changing `lr_policy: "fixed"` to `lr_policy: "step"` and adding `gamma: 0.5` and `stepsize: 50000`.
* The cosine schedule is quite fancy and the authors of [SGDR: Stochastic Gradient Descent with warm restarts](https://arxiv.org/pdf/1608.03983.pdf) could achieve quite good results with this schedule. In a nutshell, the idea is to aim for an ensemble model. The learning rate in each iteration is reduced following a function (in this case the cosine with predefined period) of the iteration. When lr 0 is reached, the snapshot is stored and the base_lr re-initialized to the initial learning rate. This is repeated a few times. Then the resulting models are applied and the majority-class assigned. 

Although this is quite cool, it is not directly applicable using caffe. But, if you want to go into these kinds of optimization, you should consider keras/tensorflow or pytorch anyways, this gives you 100% flexibility.

**Comment by jlause** [comment]
Hey Thorsten,
thanks for the throughout & quick reply! I think I'll try the step schedule then and see how far that gets me.
Cheers!
Jan.
